### PR TITLE
DP algorithm for `minimum_generating_set`

### PR DESCRIPTION
**Pull request by pranav-joshi-iitgn**
### Changes
1. Changed the recursive algorithm into one that uses Dynamic Programming that uses Chief Series, which is way faster than computing Minimal Normal Subgroups at every iteration. The algorithm can be understood with this flowchart :

MinGenSet

2. Removed the comparison of groups by checking whether all elements are equal. I'm using the `==` operator instead. Now, the `lift` subroutine takes time comparable to computation of Quotient groups (the homomorphism and its image source combined)  by GAP methods. Now, the function can compute minimum generating set of large groups like $A_5^7$ easily, while previously it could not.
3. Used the generator as much as I can in the `lift` subroutine to avoid creating new lists.
4. Relying on `libgap.MinimalGeneratingSet` where it can work.
5. Got rid of all the type conversion, which is mostly unnecessary.

### Timing Analysis
The algorithm is still a poly-time algorithm in $|G|$ . This is evident from these graphs : 

Figure 7
Figure 6
Figure 5
Figure 4
Figure 3
Figure 2
Figure 1

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.

### CODE REVIEW

File: src/sage/categories/groups.py [hunk 1]
```diff
@@ -453,51 +453,26 @@ def conjugacy_class(self, g):
...
-
-            Test that function gives an error for infinite groups::
-
-                sage: G = GL(2, ZZ)
-                sage: G.minimum_generating_set()
-                Traceback (most recent call last):
-                ...
-                NotImplementedError: only implemented for finite groups
+                sage: def A_5_to_n(n):
+                ....:     A5 = AlternatingGroup(5).gap()
```
Comment (RuchitJagodara) [inline src/sage/categories/groups.py]: Do not change anything except algorithm section in the docstring because this is also not in proper format which sagemath uses.

File: src/sage/groups/libgap_mixin.py [hunk 2]
```diff
@@ -968,132 +968,238 @@ def is_isomorphic(self, H):
         """
         return self.gap().IsomorphismGroups(H.gap()) != libgap.fail
 
-
-def minimum_generating_set(G) -> list:
-    """
-    Return a list of the minimum generating set of ``G``.
+def minimum_generating_set(G: GapElement) -> list:
+    r"""
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py:972]:
```suggestion
    """
```

File: src/sage/groups/libgap_mixin.py [hunk 3]
```diff
@@ -968,132 +968,238 @@ def is_isomorphic(self, H):
...
+    Returns the minimum generating
+    set of group G.
 
     INPUT:
 
-    - ``G`` -- a group
+    - ``G`` -- a finite group
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]:
```suggestion
    - ``G`` -- a group
```

Comment (RuchitJagodara) [reply src/sage/groups/libgap_mixin.py]: No need to mention this as in the future minimum_generating_set will be used for infinite groups also and if we constraint this then it does not define correctly the functionality of the function.

File: src/sage/groups/libgap_mixin.py [hunk 4]
```diff
@@ -968,132 +968,238 @@ def is_isomorphic(self, H):
...
 
     OUTPUT:
 
-    A list of GAP objects that generate the group.
+    - a minimum generating set of ``G``
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]:
```suggestion
    A list of GAP objects that generate the group.
```

Comment (RuchitJagodara) [reply src/sage/groups/libgap_mixin.py]: Also, I don't think we should change this as it was suggested by the maintainer of sagemath.

File: src/sage/groups/libgap_mixin.py [hunk 5]
```diff
@@ -968,132 +968,238 @@ def is_isomorphic(self, H):
...
 
     .. SEEALSO::
 
-        :meth:`sage.categories.groups.Groups.ParentMethods.minimum_generating_set`
+    :meth:`sage.categories.groups.Groups.ParentMethods.minimum_generating_set`
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]: Indentation is required here, as I told you earlier do not change anything in docstring except the algorithm section.

File: src/sage/groups/libgap_mixin.py [hunk 6]
```diff
@@ -968,132 +968,238 @@ def is_isomorphic(self, H):
...
+    The `lift` function returns the
+    (representatives of) minimum gnerating set
+    of `G/G_k` given the (representatives of)
+    minimum generating set of `G/G_{k-1}`.
+    It is discussed later when it is encountered in code.
+
+    return `g`
+
+    EXAMPLES::
+
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py:1054]: do not change anything except the algorithm section in the docstring.

**Review by RuchitJagodara — CHANGES_REQUESTED** [review]

File: src/sage/groups/libgap_mixin.py (hunk 2 above)
Comment (pranav-joshi-iitgn) [reply src/sage/groups/libgap_mixin.py:972]: There's a lot of LaTeX in the docstring, which needs a lot of `\` . I need it to be a raw string for it to work as intended.

File: src/sage/categories/groups.py [hunk 7]
```diff
@@ -451,53 +451,22 @@ def conjugacy_class(self, g):
             from sage.groups.conjugacy_classes import ConjugacyClass
             return ConjugacyClass(self, g)
 
-        def minimum_generating_set(self):
-            """
-            Return a list of the minimum generating set of this group.
+        def minimum_generating_set(self) -> list:
+            r"""
+            Returns a the minimum generating set of this group.
```
Comment (RuchitJagodara) [inline src/sage/categories/groups.py]:
```suggestion
            Returns a list of the minimum generating set of this group.
```

File: src/sage/categories/groups.py [hunk 8]
```diff
@@ -451,53 +451,22 @@ def conjugacy_class(self, g):
...
-
-            Test that function gives an error for infinite groups::
-
-                sage: G = GL(2, ZZ)
-                sage: G.minimum_generating_set()
-                Traceback (most recent call last):
-                ...
-                NotImplementedError: only implemented for finite groups
+                sage: P = PermutationGroup([(1,2,3,4,5),(3,4,5),(6,7,8,9,10),(8,9,10)]) # A group isomoprphic to A_5 ^ 2
+                sage: g = P.minimum_generating_set()
```
Comment (RuchitJagodara) [inline src/sage/categories/groups.py:461]: Why did you delete last examples which were already there just keep it there and if you want you can add new ones after that.

File: src/sage/groups/libgap_mixin.py [hunk 9]
```diff
@@ -968,10 +968,10 @@ def is_isomorphic(self, H):
         """
         return self.gap().IsomorphismGroups(H.gap()) != libgap.fail
 
-
-def minimum_generating_set(G) -> list:
-    """
-    Return a list of the minimum generating set of ``G``.
+def minimum_generating_set(G: GapElement) -> list:
+    r"""
+    Returns the minimum generating
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]:
```suggestion
    Returns a list of the minimum generating set of group ``G``
```

File: src/sage/groups/libgap_mixin.py [hunk 10]
```diff
@@ -968,10 +968,10 @@ def is_isomorphic(self, H):
...
+    set of group G.
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]:
```suggestion
```

Comment (RuchitJagodara) [reply src/sage/groups/libgap_mixin.py]: This line is not needed to be added on another line.

File: src/sage/groups/libgap_mixin.py [hunk 11]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
-    (where lg is the length of g) is able to generate G. If not, it uses the set g_1, g_2, ..., g_lg, s_j
-    as the minimum generating set of G.
+    for k = 2 to `l` ,
+    compute `G/G_k` , `G_{k-1}/G_k` and set
+    `g := lift(g, (G_{k-1}/G_k) , (G/G_k) )`
 
-    If N is non-abelian, the algorithm checks if any case of the form g_1*n_1, g_2*n_2, ..., g_lg*n_lg
-    is able to generate G, where n_1, n_2, ..., n_lg can be any elements of N. If not, it checks if any
-    case of the form g_1*n_1, g_2*n_2, ..., g_lg*n_lg, n_lg+1 is able to generate G.
+    The `lift` function returns the
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]: you can write more content in one line the limit is of 60 characters so try to add more contents on the same line so that number of line gets reduced.

File: src/sage/groups/libgap_mixin.py [hunk 12]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
-        sage: set(p.AsList()) == set(libgap.GroupByGenerators(s).AsList())
+        sage: A5 = AlternatingGroup(5).gap()
+        sage: G = A5.DirectProduct(A5)
+        sage: g = minimum_generating_set(G)
+        sage: G == libgap.GroupByGenerators(g)
         True
+        sage: len(g)
+        2
 
-    Test that elements of resultant list are GAP objects::
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py:1022]: Here also you have deleted some test cases please revert it.

File: src/sage/groups/libgap_mixin.py [hunk 13]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
-        sage: s[0].parent()
-        C library interface to GAP
     """
-    if not isinstance(G, GapElement):
-        try:
-            G = G.gap()
-        except (AttributeError, ValueError, TypeError):
-            raise NotImplementedError("only implemented for groups that can construct a gap group")
-
+    assert isinstance(G, GapElement)
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py:1064]: Instead of assert statement write a assertionerror with a error message "group G should be an instance of GAP group."

File: src/sage/groups/libgap_mixin.py [hunk 14]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
+        given the minimum generating set
+        (as representatives) of `G/G_{i-1}`,
+        the factor group `G_{i-1}/G_i` and
+        the quotient group `G/G_i` itself.
 
-    def is_group_by_gens(group, gens):
-        return set(group.AsList()) == set(libgap.GroupByGenerators(gens).AsList())
+        INPUT:
+
+        - ``G_by_Gim1_mingen_reps`` -- representative elements
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]: No need to add docstring here. Whenever you hover on a function name, there appears a dialogue box and the content of the docstring appears there but while we define a nested function (function which is inside another function we actually don't need that and because of that we don't need docstring here and also according to sage's guidelines we should not do this so just add a small comment what this function does in very short way or verbose way whatever you prefer. and remove the docstring)

Comment (RuchitJagodara) [reply src/sage/groups/libgap_mixin.py]: Also no need to describe input and output

File: src/sage/groups/libgap_mixin.py [hunk 15]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
-    if N.IsAbelian().sage():
-        if is_group_by_gens(G, g):
-            return g
+        `\bold{n} :=\{n_1,n_2\dots n_k\}`
+        where `\{n_1 G_i,n_2G_i \dots n_kG_{i}\}`
+        is any generating set of `G_{i-1}/G_i ,
+        i.e. it's the representative elements of
+        any (prefferably small,
+        but not necessarily minimal)
+        generating set of `G_{i-1}/G_i`
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]:
```suggestion
        `\bold{n} :=\{n_1,n_2\dots n_k\}`
        where `\{n_1 G_i,n_2G_i \dots n_kG_{i}\}` is any generating set of `G_{i-1}/G_i ,
        i.e. it's the representative elements of any (prefferably small, but not necessarily minimal)
        generating set of `G_{i-1}/G_i`
```

Comment (RuchitJagodara) [reply src/sage/groups/libgap_mixin.py]: Here also you were taking too much lines for just a small thing

File: src/sage/groups/libgap_mixin.py [hunk 16]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
-            for j in range(len(S)):
-                temp = g[i]
-                g[i] *= S[j]
-                if is_group_by_gens(G, g):
-                    return g
-                g[i] = temp
+        `\bold{N} := \{N_1,N_2\dots N_m\}` where
+        `G_{i-1}/G_i = \{N_1G_i,N_2G_2\dots N_m G_m\}`.
+        This is simply a list of
+        representative elements of `G_{i-1}/G_i`.
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]:
```suggestion
        `\bold{N} := \{N_1,N_2\dots N_m\}` where `G_{i-1}/G_i = \{N_1G_i,N_2G_2\dots N_m G_m\}`.
        This is simply a list of representative elements of `G_{i-1}/G_i`.
```

File: src/sage/groups/libgap_mixin.py [hunk 17]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
 
-        return g + [S[0]]
+        We wish to find the representatives of
+        a minimum generating set of `G/G_i`.
+        Here, we have two cases to consider.
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]:
```suggestion
        We wish to find the representatives of a minimum generating set of `G/G_i`.
        Here, we have two cases to consider.
```

File: src/sage/groups/libgap_mixin.py [hunk 18]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
+        Then, for all combinations of
+        (not necessarily distinct) elements
+        `N_{i_1},N_{i_2}\dots N_{i_t} N_{i_{t+1}} \in \bold{N}`,
+        Compute `g^* =`
+        `\{g_1N_{i_1},g_{i_2}N_{i_3}\dots`
+        `g_{i_t}N_t,g_{t+1}\dots g_s\}`.
+        If `\braket{g^*G_i}\; = G/G_i`, return `{g^*}`
+
+        By now, we must have exhausted our search.
+
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]: you should write this whole thing in main docstring of the function i.e. minimum_generating_set , as you can not have a docstring for this nested function and also remove this whole docstring and shift this part to the main function's docstring.

File: src/sage/groups/libgap_mixin.py [hunk 19]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
+            INPUT:
+
+                - ``g`` -- a list of elements from some group `G`
+
+                - ``N`` -- a list of elements of
+                  normal subgroup `N` of `G`
+
+                - ``t`` -- the index till which to modify ``g``
+            """
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]:
```suggestion
```

File: src/sage/groups/libgap_mixin.py [hunk 20]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
+                    gm[t-1] = old * n
+                    yield gm
+                    gm[t-1] = old
+        s = len(G_by_Gim1_mingen_reps)
+        Gim1_by_Gi_L = list(Gim1_by_Gi.AsList())
+        Gim1_by_Gi_elem_reps = [
+            phi_Gim1_by_Gi.PreImagesRepresentative(x) for x in Gim1_by_Gi_L]
+        Gim1_by_Gi_gen = list(libgap.SmallGeneratingSet(Gim1_by_Gi))
+        Gim1_by_Gi_gen_reps = [
+            phi_Gim1_by_Gi.PreImagesRepresentative(x) for x in Gim1_by_Gi_gen]
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]:
```suggestion
            if t == 0:
                yield g
                return
            for gm in gen_combinations(g, N, t-1):
                for n in N:
                    old = gm[t-1]
                    gm[t-1] = old * n
                    yield gm
                    gm[t-1] = old
   
        s = len(G_by_Gim1_mingen_reps)
        Gim1_by_Gi_L = list(Gim1_by_Gi.AsList())
    
        Gim1_by_Gi_elem_reps = [
            phi_Gim1_by_Gi.PreImagesRepresentative(x) for x in Gim1_by_Gi_L]
        Gim1_by_Gi_gen = list(libgap.SmallGeneratingSet(Gim1_by_Gi))
        Gim1_by_Gi_gen_reps = [
            phi_Gim1_by_Gi.PreImagesRepresentative(x) for x in Gim1_by_Gi_gen]
```

File: src/sage/groups/libgap_mixin.py [hunk 21]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
+                        return G_by_Gim1_mingen_reps
+                    G_by_Gim1_mingen_reps[i] = temp
+            return G_by_Gim1_mingen_reps + [Gim1_by_Gi_gen_reps[0]]
+        for raw_gens in gen_combinations(
+                G_by_Gim1_mingen_reps,
+                Gim1_by_Gi_elem_reps, s):
+            if (G_by_Gi == libgap.GroupByGenerators(
+                [phi_G_by_Gi.ImagesRepresentative(x)
+                 for x in raw_gens])):
+                return raw_gens
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]:
```suggestion
        if Gim1_by_Gi.IsAbelian().sage():
            if (G_by_Gi == libgap.GroupByGenerators(
                [phi_G_by_Gi.ImagesRepresentative(x)
                 for x in G_by_Gim1_mingen_reps])):
                return G_by_Gim1_mingen_reps
                
            for i in range(s):
                for j in range(len(Gim1_by_Gi_gen_reps)):
                    temp = G_by_Gim1_mingen_reps[i]
                    G_by_Gim1_mingen_reps[i] = G_by_Gim1_mingen_reps[i] * \
                        Gim1_by_Gi_gen_reps[j]
                        
                    if (G_by_Gi == libgap.GroupByGenerators(
                        [phi_G_by_Gi.ImagesRepresentative(x)
                         for x in G_by_Gim1_mingen_reps])):
                        return G_by_Gim1_mingen_reps
                    G_by_Gim1_mingen_reps[i] = temp
                    
            return G_by_Gim1_mingen_reps + [Gim1_by_Gi_gen_reps[0]]
            
        for raw_gens in gen_combinations(
                G_by_Gim1_mingen_reps,
                Gim1_by_Gi_elem_reps, s):
            if (G_by_Gi == libgap.GroupByGenerators(
                [phi_G_by_Gi.ImagesRepresentative(x)
                 for x in raw_gens])):
                return raw_gens
```

File: src/sage/groups/libgap_mixin.py [hunk 22]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
+        for raw_gens in gen_combinations(
+                G_by_Gim1_mingen_reps+[Gim1_by_Gi_elem_reps[0]],
+                Gim1_by_Gi_elem_reps, s+1):
+            if (G_by_Gi == libgap.GroupByGenerators(
+                [phi_G_by_Gi.ImagesRepresentative(x)
+                 for x in raw_gens])):
+                return raw_gens
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py:1133]: this is not according to the pep8 rules.

File: src/sage/groups/libgap_mixin.py [hunk 23]
```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
...
+    for k in range(2, len(cs)):
+        mingenset_km1_reps = mingenset_k_reps
+        Gk, Gkm1 = cs[k], cs[k-1]
+        phi_GbyGk = G.NaturalHomomorphismByNormalSubgroup(Gk)
+        GbyGk = phi_GbyGk.ImagesSource()
+        phi_Gkm1byGk = Gkm1.NaturalHomomorphismByNormalSubgroup(Gk)
+        Gkm1byGk = phi_Gkm1byGk.ImagesSource()
+        mingenset_k_reps = lift(
+            mingenset_km1_reps, Gkm1byGk, GbyGk, phi_GbyGk, phi_Gkm1byGk)
+    return mingenset_k_reps
```
Comment (RuchitJagodara) [inline src/sage/groups/libgap_mixin.py]:
```suggestion
    cs = G.ChiefSeries()
    phi_GbyG1 = G.NaturalHomomorphismByNormalSubgroup(cs[1])
    GbyG1 = phi_GbyG1.ImagesSource()
    mingenset_k_reps = [phi_GbyG1.PreImagesRepresentative(x) for x in list(
        libgap.SmallGeneratingSet(GbyG1))]  # k=1 initially
        
    for k in range(2, len(cs)):
        mingenset_km1_reps = mingenset_k_reps
        Gk, Gkm1 = cs[k], cs[k-1]
        phi_GbyGk = G.NaturalHomomorphismByNormalSubgroup(Gk)
        GbyGk = phi_GbyGk.ImagesSource()
        phi_Gkm1byGk = Gkm1.NaturalHomomorphismByNormalSubgroup(Gk)
        Gkm1byGk = phi_Gkm1byGk.ImagesSource()
        mingenset_k_reps = lift(
            mingenset_km1_reps, Gkm1byGk, GbyGk, phi_GbyGk, phi_Gkm1byGk)
            
    return mingenset_k_reps
```

**Review by RuchitJagodara — CHANGES_REQUESTED** [review]

**Comment by RuchitJagodara** [comment]
I'll improve the codestyle thing, but you have to remove the docstring from the lift function as I told you. And also complete the algorithm section which is inside the main function's docstring do this because after you complete then and then only I can start working on this.

File: src/sage/categories/groups.py (hunk 8 above)
Comment (pranav-joshi-iitgn) [reply src/sage/categories/groups.py:461]: The output for GL groups has changed. That's why I removed it. I also removed examples which `libgap.MinimalGeneratingSet` works on.

File: src/sage/groups/libgap_mixin.py (hunk 11 above)
Comment (pranav-joshi-iitgn) [reply src/sage/groups/libgap_mixin.py]: Ok

**Comment by RuchitJagodara** [comment]
What did you update last because it needed to be updated on my local version also...

File: src/sage/groups/libgap_mixin.py (hunk 22 above)
Comment (pranav-joshi-iitgn) [reply src/sage/groups/libgap_mixin.py:1133]: What should I change ?
